Allow creating a variable length net_buf pool with alignment restrictions

jukkar opened the pull request:
Add suitable macros to net_buf header which allows user to create properly aligned variable length net_buf pools.

The use case for this is to allow zero-copy usage for Nordic wifi chip when using variable length net_buf data.

Target for this is 4.3 release.

jukkar committed:
lib: net_buf: Add a way to create pool with aligned data

Add a new macro NET_BUF_POOL_VAR_ALIGN_DEFINE() which can be
used to create a pool with aligned data where both the start
of the data and the length of the data is aligned.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Add alignment support for net_buf data length alloc

The alignment value tells the amount of alignment of buffer length
when allocating net_buf data for sending. By default there is no
special alignment. This is needed for example with Nordic Wi-Fi
chip that uses SPI driver that expects 4 byte alignment for the
length of the data that is being sent.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jhedberg reviewed (commented):
Looks fine to me, but please fix the CI failures

jukkar commented:
> Looks fine to me, but please fix the CI failures

Yes, I will fix the CI stuff, just wanted to get ack/nak before that. This is anyway 4.3 stuff so we are not very hurry here.

jukkar committed:
tests: net_buf: Add alignment tests

Make sure that net_buf pools created by NET_BUF_POOL_VAR_ALIGN_DEFINE
are aligned according to user needs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jhedberg commented:
@jukkar I guess the section reserved for the refcount should be transformed to be a struct containing the refcount + padding that is MAX(sizeof(void *), alignment)

jfischer-no commented:
> I started to create a test for this and then realized that the data that is allocated using `k_heap_aligned_alloc()` is aligned as the user wants, but there is `sizeof(void *)` extra data (ref count) before the `net_buf->data`. In other words, the `k_heap_aligned_alloc()` returns a pointer that is stored to `net_buf->__buf` but that is `sizeof(void *)` before `net_buf->data` pointer the user uses. This means that the `net_buf->data` pointer might not be aligned according to user needs. The `sizeof(void *)` handling is all over the net_buf code so this needs more changes than originally thought.

IIRC cb->ref (struct net_buf_data_cb.ref) is only used in net_buf_clone to avoid a memcpy. There are not that many net_buf_clone() users. It would be nice to get rid of it. Or you can implement own struct net_buf_data_cb / pool_data_alloc() / pool_data_unref() to work around it. That is what the UDC buffer implements. https://github.com/zephyrproject-rtos/zephyr/blob/4852719f98b3bac9c05196d9855d897f47e3c82f/drivers/usb/udc/udc_common.c#L24-L52

The data buffers self should not be abused to store any metadata, otherwise they are not DMAable. It will be a painful process to fix it overall in Zephyr.

jukkar commented:
Updated version and tests added. I did not try to deal with the `net_buf_clone()`, possible changes/removal of it can be dealt in a separate PR.

jfischer-no commented on lib/net_buf/buf.c:
```diff
@@ -109,9 +112,26 @@ static uint8_t *mem_pool_data_alloc(struct net_buf *buf, size_t *size,
+		b = k_heap_alloc(pool, sizeof(void *) + *size, timeout);
+
+	} else {
+		if (*size < buf_pool->alloc->alignment) {
+			NET_BUF_DBG("Requested size %zu is smaller than alignment %zu",
+				    *size, buf_pool->alloc->alignment);
+			return NULL;
```
I do not think this is correct. Zero size buffers are necessary and useful for some protocol handling, for example USB uses them. Also, smaller values should be rounded up.

jfischer-no commented on lib/net_buf/buf.c:
```diff
@@ -121,7 +141,7 @@ static uint8_t *mem_pool_data_alloc(struct net_buf *buf, size_t *size,
 
 	/* Return pointer to the byte following the ref count */
-	return ref_count + sizeof(void *);
+	return ref_count + GET_ALIGN(buf_pool);
```
This is a waste of RAM on a platform where it must be aligned to the DCACHE line size. I prefer not to store any metadata in the buffer space.
